sql: protect WaitGroup decrement in CopyIn via sync.Once

yuzefovich opened the pull request:
We've recently seen "negative WaitGroup counter" server crash during COPY FROM execution a few times, but we have been unable to understand the root cause. It appears that the problem can happen right after the COPY execution is canceled due to `statement_timeout`. The synchronization setup is the following:
- the network-handling goroutine calls `wg.Add(1)`, pushes CopyIn command onto the stmt buf, and then blocks via `wg.Wait()`
- the copy-handling connExecutor calls `wg.Done()` in the defer of `execCopyIn`. It must be the case that that defer is executed at least twice, but it's unclear to me how that can happen.

In the absence of understanding of how this can happen and with no reproduction, this commit attempts to mitigate the problem by ensuring that `wg.Done()` is called exactly once. This is achieved via `sync.Once`.

Fixes: #112095.

Release note: None

yuzefovich commented:
I'm not very happy with this change, but I don't have any other ideas on how to mitigate #112095 / https://github.com/cockroachlabs/support/issues/2741. Curious about thoughts on whether it's worth making this change or not.

rafiss commented on pkg/sql/conn_executor.go:
```diff
@@ -3006,7 +3006,9 @@ func (ex *connExecutor) execCopyIn(
 
 	// When we're done, unblock the network connection.
-	defer cmd.CopyDone.Done()
+	defer func() {
```
one more idea instead of a buffered channel: could we change this to something like:
```
once.Do(func() { close(ch) })
```
or
```
once.Do(func() { cmd.CopyDone.Done() })
```

yuzefovich committed:
sql: protect WaitGroup decrement in CopyIn via sync.Once

We've recently seen "negative WaitGroup counter" server crash during
COPY FROM execution a few times, but we have been unable to understand
the root cause. It appears that the problem can happen right after the
COPY execution is canceled due to `statement_timeout`. The
synchronization setup is the following:
- the network-handling goroutine calls `wg.Add(1)`, pushes CopyIn
command onto the stmt buf, and then blocks via `wg.Wait()`
- the copy-handling connExecutor calls `wg.Done()` in the defer of
`execCopyIn`. It must be the case that that defer is executed at least
twice, but it's unclear to me how that can happen.

In the absence of understanding of how this can happen and with no
reproduction, this commit attempts to mitigate the problem by ensuring
that `wg.Done()` is called exactly once. This is achieved via
`sync.Once`.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115712)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 3009 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/115712#gh-1417880287:-Nl-hQ9n2o-r2_Ey3AjJ:b-p1fz7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/396f7b2a041f25ccd37241f4b7eeacddb4a84e4d/pkg/sql/conn_executor.go#L3009)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

one more idea instead of a buffered channel: could we change this to something like:
```
once.Do(func() { close(cmd.CopyDone) })
```
or
```
once.Do(func() { cmd.CopyDone.Done() })
```
</blockquote></details>

I like it, thanks, done.


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/conn_io.go:
```diff
@@ -370,6 +370,8 @@ type CopyIn struct {
 	// the connection is being handed back to the network routine.
 	CopyDone *sync.WaitGroup
+	// Once is used to decrement CopyDone exactly once.
+	Once *sync.Once
```
one thought i had is if this would be more clear with an embedded struct:
```
	CopyDone struct {
		*sync.WaitGroup
		*sync.Once
	}
```

(so that it's obvious the two should be used together

rafiss reviewed (approved):
lgtm. ty for fixing!

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/115712#-:-Nl4Glx64obHCbRDPPXP:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/conn_io.go` line 374 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/115712#gh-1418238698:-Nl4GkCL3QuxNs6bJiz8:b-bhrc22) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fb1d63a35ae0bc8b931648f8951768eb106d348/pkg/sql/conn_io.go#L374)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

one thought i had is if this would be more clear with an embedded struct:
```
	CopyDone struct {
		*sync.WaitGroup
		*sync.Once
	}
```

(so that it's obvious the two should be used together
</blockquote></details>

Very nice, done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
blathers backport 23.2.0

yuzefovich commented:
blathers backport 23.2.0-rc
